### PR TITLE
Break up stairway plot functionality.

### PR DESCRIPTION
**Pull request by jeromekelleher**
WIP for #4.

The idea here is to move away from scripts invoked via the command line and more towards Pythn functions invoked directly through snakemake.

I tried lots of different ways of getting snakemake to deal with the different stairwayplot files, but in the end it just didn't seem worth it. It's pretty easy to follow what's going on now, IMO, and the using concurrent.futures will work fine unless we're moving to using a cluster queueing system to schedule the theta estimation jobs.

I ran out of time while doing this, but thought I'd share this much. What do you think @andrewkern?

### CODE REVIEW

**Comment by andrewkern** [comment]
looking this over @jeromekelleher and it looks quite excellent. one small point here-- I wonder if we want to set this up for running on a cluster at this point. It strikes me that each bit here, from simulation to analysis, should probably be farmed out to an HPC. Thoughts?

**Comment by jeromekelleher** [comment]
> looking this over @jeromekelleher and it looks quite excellent. one small point here-- I wonder if we want to set this up for running on a cluster at this point. It strikes me that each bit here, from simulation to analysis, should probably be farmed out to an HPC. Thoughts?

F-strings and pathlib for the win!

WRT to cluster stuff, I think snakemake handles it [pretty well](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html?highlight=hpc#cluster-configuration). The only issue here is whether to try to have rules to generate all the stairwayplot bootstraps individually (which we can do), or to consider this part atomic. I've ended going the atomic route as they really don't take that long, and you'd end up batching a bunch of them up anyway on a cluster submission as it'd be really inefficient otherwise.

What's the longest you think one of these individual bootstraps would take? What's the largest number of bootstraps you can imagine us using?

**Comment by jeromekelleher** [comment]
OK, here's an updated tidied up version. Should work with, e.g.,
```
snakemake --cores=4
```

**Comment by andrewkern** [comment]
i think leaving all the bootstraps together is fine. they are quite fast on the simulations we are doing here (chr22). Moreover, I'm not even sure that we will use bootstraps when it comes time to make figures for the paper as what we will probably want is point estimates from a number of replicates, rather than bootstrap uncertainty estimates from a single point estimate.

**Comment by jeromekelleher** [comment]
> i think leaving all the bootstraps together is fine. they are quite fast on the simulations we are doing here (chr22). Moreover, I'm not even sure that we will use bootstraps when it comes time to make figures for the paper as what we will probably want is point estimates from a number of replicates, rather than bootstrap uncertainty estimates from a single point estimate.

SGTM. This is ready for review and merging now anyway, I think it's a good base to build from. Snakemake is very nice, thanks for bringing it in!

**Comment by andrewkern** [comment]
I just tagged @agladstein and @jgallowa07 to help on review here. I will have a bit of time myself too, so if we get two reviews that's plenty.

File: n_t/stairway.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        """
+        num_runs = 1
+        dim_factor = 5000
+        cmd = (
+            f"{self.java_exe} -cp {self.classpath} Stairway_plot_theta_estimation02 "
+            f"{input_file} {num_runs} {dim_factor}")
+        logging.info("Running:" + cmd)
+        subprocess.run(cmd, shell=True, check=True)
+
+    def run_theta_estimation(self, max_workers=None, show_progress=False):
```
Comment (andrewkern) [inline n_t/stairway.py]: this is threading is fucking awesome @jeromekelleher!

**Comment by andrewkern** [comment]
code looks great and ran fast using the multithreading that you added @jeromekelleher. my only note is that the threading adds a dependency to `tqdm` that we should note in `requirements.txt`

File: n_t/stairway.py (hunk 1 above)
Comment (jeromekelleher) [reply n_t/stairway.py]: Thanks @andrewkern, my feelings entirely! Loving the freedom to use recent Python3 tools and tqdm is life-changing.

Comment (jeromekelleher) [reply n_t/stairway.py]: Oh yeah, I've added tqdm to requirements.txt

**Comment by agladstein** [comment]
> I just tagged @agladstein and @jgallowa07 to help on review here. I will have a bit of time myself too, so if we get two reviews that's plenty.

I probably wan't get a chance to check this until next week.

**Comment by andrewkern** [comment]
no prob @agladstein -- i'll lean on @jgallowa07 to get the second review done. the code is clean and works so shouldn't be an issue

**Pull request merged**